Add PKC JWKSet reloading to JWT authentication doc (#88692)
elastic/elasticsearch: x-pack/docs/en/security/authentication/jwt-realm.asciidoc

----
{"username":"user123_runas","roles":["jwt_role1"],"full_name":null,"email":null,"metadata":{},
"enabled":true,"authentication_realm":{"name":"jwt2","type":"jwt"},"lookup_realm":{"name":"native",
"type":"native"},"authentication_type":"realm"}%
----

[[jwt-realm-jwkset-reloading]]
===== PKC JWKS reloading
JWT authentication supports signature verification using PKC (Public Key Cryptography) or HMAC algorithms.

> PKC JWKS can contain public RSA and EC keys. HMAC JWKS or HMAC UTF-8 JWK contain secret keys.

Reviewer: ```suggestion
PKC JSON Web Token Key Sets (JWKS) can contain public RSA and EC keys. HMAC JWKS or an HMAC UTF-8 JWK contain secret keys.
```

Reviewer: We should define JWKS here before using them. The only other definition is in the description for `pkc_jwkset_path`.

> JWT issuers typically rotate PKC JWKs more frequently (ex: daily), because RSA and EC public keys are

Reviewer: ```suggestion
JWT issuers typically rotate PKC JWKS more frequently (such as daily), because RSA and EC public keys are
```

designed to be easier to distribute than secret keys like HMAC.

> JWT realms load a PKC JWKS, and a HMAC JWKS or HMAC UTF-8 JWK, at startup.

Reviewer: ```suggestion
JWT realms load a PKC JWKS, and an HMAC JWKS or HMAC UTF-8 JWK, at startup.
```

JWT realms can also reload PKC JWKS contents at runtime; reload is triggered by signature validation failures.

> CAUTION: HMAC JWKS or HMAC UTF-8 JWK reloading is not supported at this time.

Reviewer: ```suggestion
NOTE: HMAC JWKS or HMAC UTF-8 JWK reloading is not supported at this time.
```

Reviewer: I think that this can just be a `NOTE`. A `CAUTION` admonition is typically reserved for severe instances, such as when data loss can occur.

Load failures, parse errors, and configuration errors prevent a node from starting (and restarting).
However, runtime PKC reload errors and recoveries are handled gracefully.

IMPORTANT: All other JWT realm validations are checked before a signature failure can trigger a
PKC JWKS reload. Client authentication is strongly recommended. Narrow scoping of allowed algorithms,
allowed issuer, allowed audiences, and clock skew are all strongly recommended too. That way only trusted
client applications and realm-specific user JWT will be able to trigger PKC reload attempts.

If multiple JWT authentication signature failures happen at the same time within an {es} node, reloads are
combined to reduce the reloads sent externally.

IMPORTANT: If JWT signature failures trigger PKC JWKS reloads in different {es} nodes, reload
requests cannot be combined.

IMPORTANT: If JWT signature failures trigger PKC JWKS reloads in the same {es} node, but at different

> times, separate reload requests will not be combined.

Reviewer: I think that we can reorder this information and split it into two parts to make it flow a bit better:

### First part
> All other JWT realm validations are checked before a signature failure can 
> trigger a PKC JWKS reload. If multiple JWT authentication signature failures
> occur simultaneously with a single {es} node, reloads are combined to reduce
> the reloads that are sent externally.
> 
> Separate reload requests cannot be combined in these instances:
> 
> * If JWT signature failures trigger PKC JWKS reloads in _different_ {es} nodes
> * If JWT signature failures trigger PKC JWKS reloads in the same {es} node, but 
> at different times

### Second part
This statement is confusing me a bit; I'm not entirely sure what it's referring to:

>That way only trusted client applications and realm-specific user JWT will be able to trigger PKC reload attempts.

I'm going to take an educated guess and suggest:

> [IMPORTANT]
> ====
> Enabling client authentication is strongly recommended. Only trusted client applications and realm-specific JWT users can trigger PKC reload attempts. Additionally, configuring the following <<ref-jwt-settings,JWT security settings>> is recommended:
> 
> * `allowed_issuer`
> * `allowed_audiences`
> * `allowed_clock_skew`
> * `allowed_signature_algorithms`
> ====

Author: LGTM.

If it helps, a more complete view of the order of how settings are used in the code is currently something like this below. Adding complete details would make things too complicated, so it is a balancing act to decide how much detail to put in. The 4 settings you listed above fall into 3-6, and in slightly different order. I don't know if that matters for that you want to achieve, so I will leave it up to you.

1. client_authentication.type (with client_authentication.shared_secret, if type is shared_secret)
2. jwt.cache.size (with jwt.cache.ttl and allowed_clock_skew, if cache is enabled)
3. allowed_issuer
4. allowed_audiences
5. allowed_signature_algorithms
6. allowed_clock_skew
7. validate signature (with allowed_signature_algorithms, and one of pkc_jwkset_path, hmac_jwkset, or hmac_key)
8. if-and-only-if JWT alg is RSA or EC, try reload (or wait if reload is in progress)
9. if-and-only-if reload succeeds and new contents are different, revalidate signature

Reviewer: Thanks @justincr-elastic -- I applied these changes in b5a996149ea6994aef925ba2833b86cc45ac7fe7.
